core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java

```diff
@@ -0,0 +1,128 @@
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class BaseUpdateSnapshotRefs implements UpdateSnapshotRefs {
+
+  private final TableOperations ops;
+  private final TableMetadata base;
+  private final Map<String, SnapshotRef> refs;
+
+  BaseUpdateSnapshotRefs(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.refs = Maps.newHashMap(base.refs());
+  }
+
+  @Override
+  public UpdateSnapshotRefs tag(String name, long snapshotId) {
+    Preconditions.checkArgument(name != null, "Tag name must not be null");
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null, "Cannot find snapshot with ID: %s", snapshotId);
+    Preconditions.checkArgument(!refs.containsKey(name), "Cannot tag snapshot, ref already exists: %s", name);
+
+    refs.put(name, SnapshotRef.builderForTag(snapshotId).build());
+    return this;
+  }
+
+  @Override
+  public UpdateSnapshotRefs branch(String name, long snapshotId) {
+    Preconditions.checkArgument(name != null, "Branch name must not be null");
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null, "Cannot find snapshot with ID: %s", snapshotId);
+    Preconditions.checkArgument(!refs.containsKey(name), "Cannot create branch, ref already exists: %s", name);
+
+    refs.put(name, SnapshotRef.builderForBranch(snapshotId).build());
+    return this;
+  }
+
+  @Override
+  public UpdateSnapshotRefs remove(String name) {
+    Preconditions.checkArgument(name != null, "Ref name must not be null");
+    Preconditions.checkArgument(refs.containsKey(name), "Cannot find ref to remove: %s", name);
+    Preconditions.checkArgument(!SnapshotRef.MAIN_BRANCH.equals(name), "Main branch must not be removed");
+
+    refs.remove(name);
+    return this;
+  }
+
+  @Override
+  public UpdateSnapshotRefs rename(String from, String to) {
+    Preconditions.checkArgument(from != null && to != null, "Names must not be null");
+    Preconditions.checkArgument(refs.containsKey(from), "Cannot find ref to rename from: %s", from);
+    Preconditions.checkArgument(!refs.containsKey(to), "Cannot rename to an existing ref: %s", to);
+
+    refs.put(to, refs.remove(from));
+    return this;
+  }
+
+  @Override
+  public UpdateSnapshotRefs setLifetime(String name, long ageMs) {
+    Preconditions.checkArgument(name != null, "Branch name must not be null");
+    Preconditions.checkArgument(ageMs > 0, "Lifetime must be positive");
+    Preconditions.checkArgument(refs.containsKey(name), "Cannot find ref to set lifetime: %s", name);
+    Preconditions.checkArgument(!SnapshotRef.MAIN_BRANCH.equals(name), "Main branch is retained forever");
+
+    SnapshotRef oldRef = refs.get(name);
+    refs.put(name, SnapshotRef.builderFrom(oldRef).maxRefAgeMs(ageMs).build());
+    return this;
+  }
```

Comment:
I think maybe we should throw error when user update several properties in one operation. Because there is just one property will be update.